### ISSUE TITLE
flag nil checks followed by len check, e.g. x != nil && len(x) ... 

### ISSUE THREAD

**Issue by client9** [golang/lint#195]
This is a follow up from [golang-nuts](https://groups.google.com/forum/#!topic/golang-nuts/AYFi0CQBsNM) and a [golang ticket](https://github.com/golang/go/issues/14733).  The conclusion from both is that golint is the right place to discuss this.

The proposal is to log on the following "redundant nil checks"

```go
if x == nil || len(x) == 0 {}
if x != nil && len(x) ... {  // or any operator len(x) > 0, len(x) != 0, len(x) > 10000
```

**in most cases** can be simplified into just  

```go
if len(x) == 0 {}
if len(x) .... {}
```

@griesemer  pointed out that there are cases where this is wrong, the value is pointer to a slice/array http://play.golang.org/p/PMfEn4oT_i

```go
package main
import "fmt"
func main() {
	var a *[10]int
	fmt.Println(a, len(a))
	fmt.Println(a == nil || len(a) == 0, len(a) == 0)
}
// true, false
```

That said I think this is a bit unusual in my experience.  But perhaps my view is limited.

### How common is this ?

Not uncommon:  While this code search on GitHub isn't perfect, it does show up.  I'd prefer not to download github to correctly search until I understand the acceptance requirements (defects per 10k? at what level?  number of repos affected?)

* https://github.com/search?utf8=✓&q=%22%3D%3D+nil+%7C%7C+len%28%22+extension%3Ago&type=Code&ref=searchresults
* https://github.com/search?utf8=✓&q=%22%21%3D+nil+%26%26+len%28%22+extension%3Ago&type=Code&ref=searchresults

To quote the original topic:

> As a quick test I used grep, and removed false positives.
> 
> $ find . -name '*.go' | xargs grep '= nil || len(' | grep Godeps
> 
> ./Godeps/_workspace/src/github.com/gocql/gocql/marshal.go:	if data == nil || len(data) == 0 {
> ./Godeps/_workspace/src/github.com/pborman/uuid/uuid.go:	if uuid == nil || len(uuid) != 16 {
> ./Godeps/_workspace/src/github.com/pborman/uuid/uuid.go:	if uuid == nil || len(uuid) != 16 {
> ./Godeps/_workspace/src/github.com/ugorji/go/codec/decode.go:			if rvbs == nil && bs2 != nil || rvbs != nil && bs2 == nil || len(bs2) != len(rvbs) { ###  wow!!
>
>$ find . -name '*.go' | xargs grep '= nil && len(' | grep Godeps
>
> ./Godeps/_workspace/src/github.com/araddon/gou/jsonhelper.go:if lst := jsonList(root); lst != nil && len(lst) > idx {
> ./Godeps/_workspace/src/github.com//elastigo/lib/clusterhealth.go:	if f.FilterIndices != nil && len(f.FilterIndices) > 0 {
>
> thats 5 repos out of 60.  Granted a small sample but that doesn't seem so rare.

I reran this, a few of them have been fixed up and some new ones added.

Again, this isn't taking into account the pointer to array situation. 

However I think the lint error is still ok even with the "pointer to array" case. 

The proposal would only take into account where a variable was checked for nil FIRST, then the len.  So if we have a case like http://play.golang.org/p/PMfEn4oT_i  it could be rewritten as

```go
len(a) == 0 || a == nil
```

to avoid being beeped.

I'm open to ideas here.  Or if people think it's a bad check, thats ok too!

regards,

nickg

**Comment by dsymonds** [comment]
I agree golint would be the right place for such a check.

I'm not sure it's a particularly worthwhile check, though. It's not super common, as your data suggests, and it doesn't feel particularly egregious from a stylistic perspective. The only redeeming aspect of it is that it is machine checkable, but that's quite a lot bar.

I'm open to further arguments, but I'm leaning towards "no" right now.

**Comment by client9** [comment]
Hi, thanks for taking a look.

It's tough to make arguments since I'm not sure I understand acceptance criteria for golint.  Is it just for stuff referenced in Effective Go and stuff in CodeReviewComments?   If so, then yeah, this check doesn't match anything there, and I should not have even filed a ticket. To make it more clear,  I would move last sentence in the Purpose section of the README to the top.  It's kinda the buried lead.

If golint is more general "best practice", then almost any change would not be very common else we've been all doing it wrong for a long time :-)

My argument is:  for newcomers to go, the redundant check is a natural mistake to make since other languages make this check NOT redundant (i.e. it's required).  The golint warning would be a nice reminder or education tip to show folks that go is different.  My original Go code had this all over the place (from a C/C++ background), and I would have welcomed this warning. 

BUT...that may not be purpose of golint, and very ok with that.

Again thanks for your time!

**Comment by dsymonds** [comment]
The acceptance criteria is not well defined. It amounts to things that are stylistic, reasonably common, machine checkable accurately, and actually problematic in some way if violated.

There's lots of mistakes that Go newcomers will make. I don't think golint is the place to catch them.

**Comment by wjkohnen** [comment]
I would like to see this check _somewhere_, at least for educational purposes. The benefits of having this check available somewhat reminds me of the benefits of [unconvert](https://github.com/mdempsky/unconvert). If this check can't find a home, maybe make it a standalone command as well and I am sure @alecthomas would include it in his [gometalinter](https://github.com/alecthomas/gometalinter)?

**Comment by kostya-sh** [comment]
@client9, maybe this check can be added to https://github.com/dominikh/go-simple?
